engines/ultima/ultima8/graphics/render_surface.cpp

```diff
@@ -324,10 +292,131 @@ bool RenderSurface::IsFlipped() const {
 // Desc: Fill buffer (using a RGB colour)
 //
 void RenderSurface::Fill32(uint32 rgb, const Rect &r) {
-	Rect rect = r;
+	Common::Rect rect(r.left, r.top, r.right, r.bottom);
 	rect.clip(_clipWindow);
+	rect.translate(_ox, _oy);
 	rgb = _surface->format.RGBToColor((rgb >> 16) & 0xFF, (rgb >> 8) & 0xFF, rgb & 0xFF);
-	_surface->fillRect(Common::Rect(rect.left + _ox, rect.top + _oy, rect.right + _ox, rect.bottom + _oy), rgb);
+	_surface->fillRect(rect, rgb);
+}
+
+
+// #define CHECK_ALPHA_FILLS
+namespace {
+
+template<typename uintX>
+void inline fillAlphaLogic(uint8 *pixels, int32 pitch, uint8 alpha, const Common::Rect &rect, const Graphics::PixelFormat &format) {
+	uint32 aMask = format.aMax() << format.aShift;
+	int32 w = rect.width();
+	int32 h = rect.height();
+
+	if (!w || !h || !aMask)
+		return;
+
+	uint32 a = (((uint32)alpha) << format.aShift) & aMask;
+
+#ifdef CHECK_ALPHA_FILLS
+	uint32 c;
+	uint32 m;
+	if (a == 0) {
+		uint32 bMask = format.bMax() << format.bShift;
+		c = (bMask >> 1) & bMask;
+		m = bMask;
+	} else {
+		uint32 rMask = format.rMax() << format.rShift;
+		c = (rMask >> 1) & rMask;
+		m = rMask;
+	}
+#endif
+
+	uint8 *pixel = pixels + rect.top * pitch + rect.left * format.bytesPerPixel;
+	int diff = pitch - w * format.bytesPerPixel;
+
+	for (int y = 0; y < h; ++y) {
+		for (int x = 0; x < w; ++x) {
+			uintX *dest = reinterpret_cast<uintX *>(pixel);
+			*dest = (*dest & ~aMask) | a;
+#ifdef CHECK_ALPHA_FILLS
+			*dest = (*dest & ~m) | (c + (((*dest & m) >> 1) & m));
+#endif
+			pixel += format.bytesPerPixel;
+		}
+
+		pixel += diff;
+	}
+}
+
+} // End of anonymous namespace
+
+//
+// RenderSurface::FillAlpha(uint8 alpha, Rect r)
+//
+// Desc: Fill alpha channel
+//
+void RenderSurface::FillAlpha(uint8 alpha, const Rect &r) {
```

Comment:
Am I correct in thinking that this is essentially the same as `Graphics::Surface::setAlpha()` except with an additional Rect parameter?

Reply:
I believe you are correct. It was used in Pentagram for a two pass rendering of TextWidget for the normal scaled paint and then the composited unscaled paint to render a higher res font.

I've been maintaining this method and testing with CHECK_ALPHA_FILLS, although it's currently not used as the composited painting was abandoned when ported to ScummVM.